File: src/main/java/com/fish/service/impl/AccountServiceImpl.java
package com.fish.service.impl;

import com.fish.common.Result;
import com.fish.entity.dto.AccountDTO;
import com.fish.entity.pojo.Account;
import com.fish.entity.pojo.Role;
import com.fish.entity.vo.AccountVO;
import com.fish.exception.ServiceException;
import com.fish.exception.ServiceExceptionEnum;
import com.fish.mapper.AccountMapper;
import com.fish.mapper.RoleMapper;
import com.fish.security.LoginAccount;
import com.fish.security.SystemAuthority;
import com.fish.service.AccountService;
import com.fish.utils.JwtUtil;
import com.fish.utils.RedisUtil;
import com.fish.utils.ResultUtil;
import com.mybatisflex.spring.service.impl.ServiceImpl;
import jakarta.annotation.Resource;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContext;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Objects;

@Service
public class AccountServiceImpl extends ServiceImpl<AccountMapper, Account> implements AccountService {
    @Resource
    private AuthenticationManager authenticationManager;
    @Resource
    private AccountMapper accountMapper;
    @Resource
    private RoleMapper roleMapper;
    @Resource
    private PasswordEncoder passwordEncoder;
    @Resource
    private JwtUtil jwtUtil;
    @Resource
    private RedisUtil redisUtil;
    @Value("${jwt.year}")
    private String year;
    private static final SecurityContext context = SecurityContextHolder.getContext();
    @Override
    public Result<HashMap<String,Object>> login(AccountDTO accountDTO) {
        if (checkInformation(accountDTO.getEmail(), accountDTO.getPassword(), null))
            throw new ServiceException(ServiceExceptionEnum.KEY_ARGUMENT_NOT_INPUT);
        UsernamePasswordAuthenticationToken authenticationToken =
                new UsernamePasswordAuthenticationToken(accountDTO.getEmail(), accountDTO.getPassword());
        Authentication authentication = authenticationManager.authenticate(authenticationToken);
        if (Objects.isNull(authentication)) {
            throw new ServiceException(ServiceExceptionEnum.ACCOUNT_NOT_FOUND);
        }
        LoginAccount principal = (LoginAccount) authentication.getPrincipal();
        String redisKey = makeKey(principal.getAccount().getUserId(), principal.getAccount().getNickname());
        String token = jwtUtil.createToken(redisKey);
        HashMap<String, Object> map = new HashMap<>();
        map.put("token", token);
        map.put("accountInfo", new AccountVO(principal.getAccount()));
        redisUtil.set(redisKey, principal);
        return ResultUtil.success(map);
    }
    @Transactional
    @Override
    public Result<?> register(AccountDTO accountDTO) {
        if (checkInformation(accountDTO.getEmail(), accountDTO.getPassword(), accountDTO.getNickname()))
            throw new ServiceException(ServiceExceptionEnum.KEY_ARGUMENT_NOT_INPUT);
        accountDTO.setPassword(passwordEncoder.encode(accountDTO.getPassword()));
        int i = accountMapper.addAccount(accountDTO);
        if (i > 0)
            return ResultUtil.success();
        else
            throw new ServiceException(ServiceExceptionEnum.ACCOUNT_EXISTED);
    }
    @Transactional
    @Override
    public Result<?> updateAccountInformation(AccountDTO accountDTO) {
        if (Objects.isNull(accountDTO.getUserId()))
            throw new ServiceException(ServiceExceptionEnum.KEY_ARGUMENT_NOT_INPUT);
        String credentials = (String) SecurityContextHolder.getContext().getAuthentication().getCredentials();
        if (!credentials.equals(accountDTO.getUserId()))
            throw new ServiceException(ServiceExceptionEnum.AUTHENTICATION_FAILURE);
        int i = accountMapper.updateAccountInformation(accountDTO);
        if (i > 0)
            return ResultUtil.success();
        else
            throw new ServiceException(ServiceExceptionEnum.OPERATE_ERROR);
    }
    @Transactional
    @Override
    public Result<?> deleteAccount(String userId) {
        if (checkRole(userId))
            throw new ServiceException(ServiceExceptionEnum.INSUFFICIENT_PERMISSIONS);
        int i = accountMapper.deleteAccount(userId);
        if (i > 0)
            return ResultUtil.success();
        else
            throw new ServiceException(ServiceExceptionEnum.OPERATE_ERROR);
    }
    @Override
    public Result<ArrayList<Account>> getAccounts() {
        return ResultUtil.success(accountMapper.getCustomAccounts());
    }
    /**
     * <div>先检查当前用户是否有权限修改权限信息再进行修改</div>
     * <div>
     *     比如：
     *      <div>1. ROOT 能修改 ADMIN、USER</div>
     *      <div>2. ADMIN 能修改 USER</div>
     *      <div>3. USER 则无修改权限</div>
     * </div>
     * @param userId 要修改的用户id
     * @param roleId 要修改成的角色id
     * @return 响应操作结果
     */
    @Transactional
    @Override
    public Result<?> changeAccountRole(String userId, Long roleId) {
        if (checkRole(userId) || checkRole(roleId))
            throw new ServiceException(ServiceExceptionEnum.INSUFFICIENT_PERMISSIONS);
        int i = accountMapper.changeAccountRole(userId, roleId);
        if (i > 0)
            return ResultUtil.success();
        else
            throw new ServiceException(ServiceExceptionEnum.OPERATE_ERROR);
    }
    protected boolean checkInformation(String email, String password, String nickname) {
        if (Objects.isNull(nickname))
            return (!StringUtils.hasLength(email) && !StringUtils.hasLength(password));
        else
            return (!StringUtils.hasLength(email) && !StringUtils.hasLength(password) && !StringUtils.hasLength(nickname));
    }
    protected String makeKey(String userId, String nickname) {
        return nickname + "_" + userId + "_" + year;
    }
    /**
     * 判断执行修改的用户是否足够权限更改对应权限
     * @param roleId 角色id
     * @return 是否有足够权限
     */
    protected boolean checkRole(Long roleId) {
        if (roleId == 3)
            return true;
        Authentication authentication = context.getAuthentication();
        ArrayList<SystemAuthority> authorities = new ArrayList<>();
        authentication.getAuthorities().forEach(grantedAuthority -> {authorities.add((SystemAuthority) grantedAuthority);});
        String authority = authorities.get(0).getAuthority();
        Role role = roleMapper.getRole(authority);
        return role.getRoleId() > roleId;
    }
    /**
     * <div>判断当前用户是否有权修改该用户的权限</div>
     * @param userId 用户id
     * @return 是否有足够权限
     */
    protected boolean checkRole(String userId) {
        Authentication authentication = context.getAuthentication();
        ArrayList<SystemAuthority> authorities = new ArrayList<>();
        authentication.getAuthorities().forEach(grantedAuthority -> {authorities.add((SystemAuthority) grantedAuthority);});
        String authority = authorities.get(0).getAuthority();
        Role role = roleMapper.getRole(authority);
        Account account = accountMapper.getAccount(userId);
        return role.getRoleId() > account.getRoleId();
    }
}

File: src/main/kotlin/com/fish/kexieOnlineJudge/service/account/impl/AccountServiceImpl.kt
package com.fish.kexieOnlineJudge.service.account.impl

import com.fish.kexieOnlineJudge.common.Result
import com.fish.kexieOnlineJudge.entity.dto.account.LoginAccountDTO
import com.fish.kexieOnlineJudge.entity.dto.account.PasswordDTO
import com.fish.kexieOnlineJudge.entity.dto.account.RegisterAccountDTO
import com.fish.kexieOnlineJudge.entity.dto.account.UpdateAccountDTO
import com.fish.kexieOnlineJudge.entity.pojo.account.Account
import com.fish.kexieOnlineJudge.entity.pojo.account.table.AccountTableDef.ACCOUNT
import com.fish.kexieOnlineJudge.entity.vo.AccountVO
import com.fish.kexieOnlineJudge.exception.ServiceException
import com.fish.kexieOnlineJudge.exception.ServiceExceptionEnum
import com.fish.kexieOnlineJudge.mapper.account.AccountMapper
import com.fish.kexieOnlineJudge.security.LoginAccount
import com.fish.kexieOnlineJudge.service.account.AccountService
import com.fish.kexieOnlineJudge.utils.RedisUtil
import com.fish.kexieOnlineJudge.utils.ResultUtil.success
import com.fish.kexieOnlineJudge.utils.SecurityUtil
import com.mybatisflex.core.paginate.Page
import com.mybatisflex.core.query.QueryWrapper
import com.mybatisflex.core.update.UpdateChain
import com.mybatisflex.kotlin.extensions.kproperty.column
import com.mybatisflex.kotlin.extensions.kproperty.eq
import com.mybatisflex.spring.service.impl.ServiceImpl
import org.springframework.beans.factory.annotation.Value
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class AccountServiceImpl(
    val authenticationManager: AuthenticationManager,
    val passwordEncoder: PasswordEncoder,
    val jwtUtil: com.fish.kexieOnlineJudge.utils.JwtUtil,
    val redisUtil: RedisUtil
) : ServiceImpl<AccountMapper, Account>(), AccountService {
    @Value("\${jwt.year}")
    private val year: String? = null
    override fun login(loginAccountDTO: LoginAccountDTO): Result<HashMap<String, Any>> {
        val authenticationToken = UsernamePasswordAuthenticationToken(loginAccountDTO.email, loginAccountDTO.password)
        val authentication = authenticationManager.authenticate(authenticationToken)
        authentication?.let {
            val principal = authentication.principal as LoginAccount
            val redisKey = makeKey(principal.account.userId!!, principal.account.nickname!!)
            val token = jwtUtil.createToken(redisKey)
            val map = HashMap<String, Any>()
            map["Authorization"] = token
            map["account"] = AccountVO(principal.account)
            redisUtil.set(redisKey, principal)
            return success(map)
        }
        throw ServiceException(ServiceExceptionEnum.ACCOUNT_NOT_FOUND)
    }

    @Transactional
    override fun register(registerAccountDTO: RegisterAccountDTO): Result<*> {
        registerAccountDTO.password = passwordEncoder.encode(registerAccountDTO.password)
        mapper!!.addAccount(registerAccountDTO)
        return success<Any>()
    }

    @Transactional
    override fun updateAccountInformation(updateAccountDTO: UpdateAccountDTO): Result<*> {
        val id = SecurityUtil.getId()
        updateAccountDTO.userId = id
        val i = mapper!!.updateAccount(updateAccountDTO)
        if (i > 0)
            return success<Any>()
        throw ServiceException(ServiceExceptionEnum.OPERATE_ERROR)
    }

    override fun getAccounts(pageNo: Int, pageSize: Int): Result<Page<Account>> {
        val wrapper = QueryWrapper.create().select(ACCOUNT.USER_ID, ACCOUNT.ROLE_ID, ACCOUNT.AVATAR, ACCOUNT.USERNAME,
                ACCOUNT.STUDENT_ID, ACCOUNT.GENDER, ACCOUNT.SPECIALTY, ACCOUNT.NICKNAME,
                ACCOUNT.EMAIL, ACCOUNT.QQ, ACCOUNT.GITHUB_ADDRESS, ACCOUNT.BLOG_ADDRESS, ACCOUNT.ENABLED)
            .from(ACCOUNT)
            .where(ACCOUNT.ENABLED.eq(true)).and(ACCOUNT.ROLE_ID.eq(3))
        val accountPage = mapper.paginate(Page.of(pageNo, pageSize), wrapper)
        return success(accountPage)
    }
    override fun getAdmins(pageNo: Int, pageSize: Int): Result<Page<Account>> {
        val wrapper = QueryWrapper.create().select(ACCOUNT.USER_ID, ACCOUNT.ROLE_ID, ACCOUNT.AVATAR, ACCOUNT.USERNAME,
                ACCOUNT.STUDENT_ID, ACCOUNT.GENDER, ACCOUNT.SPECIALTY, ACCOUNT.NICKNAME,
                ACCOUNT.EMAIL, ACCOUNT.QQ, ACCOUNT.GITHUB_ADDRESS, ACCOUNT.BLOG_ADDRESS, ACCOUNT.ENABLED)
            .from(ACCOUNT)
            .where(ACCOUNT.ENABLED.eq(true)).and(ACCOUNT.ROLE_ID.lt(3))
        val accountPage = mapper.paginate(Page.of(pageNo, pageSize), wrapper)
        return success(accountPage)
    }

    @Transactional
    override fun changeAccountRole(userId: String, roleId: Long): Result<*> {
        if (roleId == 1L)
            throw ServiceException(ServiceExceptionEnum.OPERATE_ERROR)
        if (checkRole(userId, roleId))
            throw ServiceException(ServiceExceptionEnum.INSUFFICIENT_PERMISSIONS)
        val i = mapper!!.changeAccountRole(userId, roleId)
        return if (i > 0) success<Any>() else throw ServiceException(
            ServiceExceptionEnum.OPERATE_ERROR
        )
    }

    @Transactional
    override fun changeStatus(userId: String, action: Boolean): Result<*> {
        val account = mapper.selectOneById(userId)
        val loginAccount = SecurityContextHolder.getContext().authentication.principal as LoginAccount
        if (loginAccount.account.roleId!! > account.roleId!!)
            throw ServiceException(ServiceExceptionEnum.INSUFFICIENT_PERMISSIONS)
        val update = UpdateChain.of(Account::class.java)
            .set(Account::enabled.column, action)
            .where(Account::userId eq userId)
            .update()
        if (update)
            return success<Any>()
        throw ServiceException(ServiceExceptionEnum.OPERATE_ERROR)
    }

    @Transactional
    override fun changePassword(passwordDTO: PasswordDTO): Result<*> {
        val id = SecurityUtil.getId()
        val account = mapper.getAccount(id)
        val matches = passwordEncoder.matches(passwordDTO.oldPwd!!, account.password)
        if (matches) {
            val update = UpdateChain.of(Account::class.java)
                .set(Account::password.column, passwordEncoder.encode(passwordDTO.newPwd)).where(Account::userId eq id).update()
            if (update) {
                redisUtil.remove(makeKey(id, account.nickname!!))
                return success<Any>()
            }
        }
        throw ServiceException(ServiceExceptionEnum.OPERATE_ERROR)
    }

    protected fun makeKey(userId: String, nickname: String): String {
        return nickname + "_" + userId + "_" + year
    }

    /**
     * <div>判断当前用户是否有足够权限</div>
     * @param userId 用户id
     * @param roleId 角色id
     * @return 是否有足够权限
     */
    protected fun checkRole(userId: String, roleId: Long?): Boolean {
        val authentication = SecurityContextHolder.getContext().authentication
        val id = authentication.credentials as String
        val loginAccount = authentication.principal as LoginAccount
        if (id == userId)
            throw ServiceException(ServiceExceptionEnum.OPERATE_ERROR)
        val loginAccountRoleId = loginAccount.account.roleId!!
        val account = mapper!!.getAccount(userId)
        return if (Objects.isNull(roleId) && !Objects.isNull(account.roleId!!))
            loginAccountRoleId > account.roleId
        else loginAccountRoleId > roleId!!
    }
}
